### ISSUE TITLE
Make tool to convert VGM files to something Echo can play

### ISSUE THREAD

**Issue by sikthehedgehog** [sikthehedgehog/Echo#11]
Because not having composer-friendly tools is just outright insulting and not tolerable in any form. I'm surprised people are more angry at Deflemask (which at least works) than at this.

**Comment by minerscale** [comment]
If the code isn't too spaghetti I wouldn't mind helping with the development of this tool. I need it asap.

**Comment by marcb0t** [comment]
Yeah definitely.
dmf2esf is really not very good at converting certain things, and does not support a lot of Deflemask effects at the moment.

**Comment by sikthehedgehog** [comment]
Practically everybody just wants VGM and nothing else so I guess I'll have to turn this into the main method of composing anything in Echo (・＿・) (though I guess the Japanese homebrewers may still be interested on the MML tool)

**Comment by minerscale** [comment]
Well, Deflemask, VGM. Doesn't really make a difference to me, just so as that there is a reliable tool that works with as many Deflemask effects as absolutely possible

**Comment by marcb0t** [comment]
I agree, with what Minerscale says on this one.

VGM is a really chunky and inefficient format. It would be better to have a tool that can fully utilize all the same effects from Deflemask, but use as little data as possible per file.

**Comment by sikthehedgehog** [comment]
> Well, Deflemask, VGM. Doesn't really make a difference to me, just so as that there is a reliable tool that works with as many Deflemask effects as absolutely possible

In my experience, ask for music and composers will just give you a VGM file, then get defensive if you ever dare to ask for the source file to work from that. Which is why a tool that converts from VGM is a must, despite the fact it's one of the least reversible formats available (the only thing even less reversible would be prerendered music like MP3).

Even this isn't really optimal since some stuff isn't possible to reverse with heuristics. I have no way to tell apart what's a PSG instrument envelope from just normal volume changes, and all frequency slides have been already decomposed into many events so that can't be undone even if I were to implement #24. And let's not get started with PCM (VGMs spit out by some exporters will have the PCM blocks properly labeled as shown by vgmplay, but if that's not the case then we're screwed).

Working from VGM is the worst idea ever, the problem is that composers are convinced that's the universal format everybody can use and can't be kicked out of it. Not helped by the fact some people made actual VGM players on hardware.

**Comment by marcb0t** [comment]
> In my experience, ask for music and composers will just give you a VGM file, then get defensive if you ever dare to ask for the source file to work from that. 

Not true. I am one of those musicians who have no problem giving a source file to a trusted friend, or programmer. In fact, me and my musician friends don't have a problem with exchanging source files with each others simply for viewing. Sometimes, we even enhance, or add to each others work without concern.

Perhaps there are some musicians who are more "finicky" about such things, but as for me and my friends, there is trust.

>  the problem is that composers are convinced that's the universal format everybody can use and can't be kicked out of it

I can't speak for anyone else, but actually, I'm a composer, and I really wish for a better format than VGM. For game production, Echo is a better option. Echo is more effecient for processing, and uses far less data on ROM per song.

See, don't even worry about this VGM nonsense. Composers mainly use Deflemask for genesis music these days anyway.

What would make more sense is to help optimize the "dmf2esf.exe" conversion tool. In it's current state, it has some bugs, and hardly supports any effects. Make it to where it can eventually support any effect that Deflemask can do straight off the tracker. VGM is really an unnecessary "middle man".

I want a full featured conversion utility that can convert straight from Deflemask to Echo. That is better for composers making music for actual genesis games hands down. You are hearing this straight from a composer who knows what he's talking about. ;)

**Comment by marcb0t** [comment]
> Probably before worrying about anything else it'd be better to update dmf2esf.

I would have to treat you out to a steak dinner if you did this! :D   

Definitely, because I like the idea and concept of "dmf2esf". Great tool to have, once all the bugs are fixed, and more features are supported (especially vibrato and LFO).

Thank you again for your support and consideration in resolving these matters!

**Comment by sikthehedgehog** [comment]
Note that I didn't make dmf2esf, you'd have to ask whoever is maintaining it :P (the only two "something2esf" tools I made are midi2esf and mml2esf)

**Comment by marcb0t** [comment]
Yes, I am aware of that. I understand that someone else produced that conversion module.

However, I was suggesting that helping that other programmer to polish up "dmf2esf" would seem a better idea than another VGM converter.  I admit that I was not clear enough about that.

On the other hand, if you went full force on producing a VGM2ESF converter, I would also be cool with that. As long as it had good support for all those sweet juicy Deflemask features and effects.

But hey, that's just my thoughts. Maybe I'm wrong. If you have other things to focus on, I understand.
